### PR TITLE
Fixes #137

### LINKED ISSUE #137: WM: failed $mech->follow_link(text => 'nonexistent') makes $mech->success still return true!
peteris....@gmail.com reported on Mar 31, 2008 

<pre>
<b>What steps will reproduce the problem?</b>
1. perl -MWWW::Mechanize -wle &#39;$mech = WWW::Mechanize-&gt;new;
$mech-&gt;get(q|http://reddit.com|); $mech-&gt;follow_link(text =&gt; q|this is
nonexistent!|); print $mech-&gt;success ? qq|true\n| : qq|false\n|&#39;

<b>What is the expected output? What do you see instead?</b>
Expected output is &#39;false&#39;. I see &#39;true&#39; instead.

<b>What version of the product are you using? On what operating system?</b>
I am using Mech 1.34 on Unix and Windows.

<b>Please provide any additional information below.</b>
I did not yet check the source to see why it happens this way, and as you said:
[19:57:52] &lt;alester&gt; Be careful when you say it&#39;s a &quot;bug&quot; that success
still returns true.
[19:58:01] &lt;alester&gt; To you it&#39;s unexpected.
[19:58:06] &lt;alester&gt; To someone else, it might be right.
</pre>

### Details

Imported from [Google Code issue 46](http://code.google.com/p/www-mechanize/issues/detail?id=46) via [archive](https://web.archive.org/web/20150227111857/http://code.google.com/p/www-mechanize/issues/detail?id=46)

* **Type**:      Defect
* **Date**:      Mar 31, 2008
* **Reporter**:  peteris....@gmail.com
* **Owner**:     ----
* **Priority**:  Critical
* **Status**:    Accepted
* **Labels**:    `WM`

### Comments
**petda...@gmail.com** commented on Nov 25, 2008 :

<pre>
(No comment was entered for this change.)
</pre>

* Summary : WM: failed $mech-&gt;follow_link(text =&gt; &#39;nonexistent&#39;) makes $mech-&gt;success still return true!

**pmuhlc...@gmail.com** commented on Mar 1, 2009 :

<pre>
I've written some code to find links to speeches by Presidential candidate John
McCain on the NYTimes.  The page returns results of a query on the NYTimes search
engine.  This particular page does not have a 'next' page link on it, so the
follow_link should return not successful, but it returns successful.  Also, the $mech
object should become undefined, instead it continues to contain the first page loaded
into it.  This is a serious problem.  I'll try followable links to see if that can
provide a workaround. 

My perl version:

This is perl, v5.10.0 built for x86_64-linux-gnu-thread-multi
</pre>

**petda...@gmail.com** commented on Jun 23, 2009 :

<pre>
(No comment was entered for this change.)
</pre>

* Status : Accepted
* Labels : `-Priority-Medium Priority-Critical`

**petda...@gmail.com** commented on Jul 6, 2009 :

<pre>
(No comment was entered for this change.)
</pre>

* Labels : `WM`

### PR DESCRIPTION
**Pull request by tcheukueppo**
Good day.

This PR is for an attempt to test the solution provided to this [issues](https://github.com/libwww-perl/WWW-Mechanize/issues/137).

To recap, it tries to follow a non-existent link in  `t/find_link_xhtml.t`

### CODE REVIEW

**Comment by simbabque** [comment]
Hello! Thanks for the PR. For the record, we've had a quick chat off-band and there'll be some refinement as it seems the test did in fact already exist. Mea culpa.

File: t/local/follow.t [hunk 1]
```diff
@@ -2,6 +2,7 @@ use warnings;
 use strict;
 use Test::More;
 use Test::Exception;
+use Test::Warn qw(warning_like);
```
[the remark this answers is not in the document]
Comment (tcheukueppo) [reply t/local/follow.t:5]: May I know why `Test::Warnings` is preferable to `Test::Warn`?

[the remark this answers is not in the document]
Comment (tcheukueppo) [reply t/local/follow.t:5]: I just tried `warning { ... }` from `Test::Warnings` and it turns out to not take into account `Carp::croak` and as a result the test fails.

[the remark this answers is not in the document]
Comment (haarg) [reply t/local/follow.t:5]: I had missed that this was using the `carped` feature. I think that's really a misfeature though. I'd say the error can be checked using Test::Warnings by doing:
```
like warning { $agent->follow_link( n => 'all' ) }, qr/^follow_link\(.*?\) is not valid/, "Can we follow all links?";
```

**Review by haarg — APPROVED** [review]

**Review by simbabque — APPROVED** [review]

**Comment by oalders** [comment]
Thanks, @tcheukueppo! 🎉

**Pull request merged**